src/slurm_plugin/common.py

```diff
@@ -8,5 +8,183 @@
+            )
+        except subprocess.CalledProcessError:
+            log.error(
+                "Encountered error when updating node %s with instance=%s private_ip=%s nodehostname=%s",
+                slurm_nodes,
+                instance_ids,
+                instance_ips,
+                instance_hostnames,
+            )
+            self.failed_nodes.extend(slurm_nodes)
+
+    @staticmethod
+    def _parse_launched_instances(launched_instances):
+        """Parse run_instance output."""
+        instance_ids = []
+        instance_ips = []
+        instance_hostnames = []
+        for instance in launched_instances:
+            instance_ids.append(instance["InstanceId"])
+            instance_ips.append(instance["PrivateIpAddress"])
+            instance_hostnames.append(instance["PrivateDnsName"].split(".")[0])
+        return instance_ids, instance_ips, instance_hostnames
+
+    def _parse_requested_instances(self):
+        """
+        Parse out instance type from slurm NodeName.
+
+        Valid NodeName format: {queue_name}-{static/dynamic}-{instance_type}-{number}
+        Sample NodeName: queue1-static-c5.xlarge-2
+        """
+        instances_to_launch = collections.defaultdict(lambda: collections.defaultdict(list))
+        for node in self.node_list:
+            capture = self._capture_nodename(node)
+            if capture:
+                queue_name, instance_type = capture
+                instances_to_launch[queue_name][instance_type].append(node)
+            else:
+                log.warning("Discarding NodeName with invalid format: %s", node)
+        log.info("instances_to_launch = %s", instances_to_launch)
+
+        return instances_to_launch
+
+    @staticmethod
+    def _launch_ec2_instances(ec2_client, cluster_name, queue, instance_type, current_batch_size):
+        """Launch a batch of ec2 instances."""
+        result = ec2_client.run_instances(
+            # To-do, evaluate best effort scaling for future
+            MinCount=current_batch_size,
+            MaxCount=current_batch_size,
+            # LaunchTemplate is different for every instance type in every queue
+            # LaunchTemplate name format: {cluster_name}-{queue_name}-{instance_type}
+            # Sample LT name: hit-queue1-c5.xlarge
+            LaunchTemplate={"LaunchTemplateName": f"{cluster_name}-{queue}-{instance_type}"},
+        )
+
+        return result["Instances"]
+
+    def _capture_nodename(self, nodename):
+        """
+        Capture queue_name and instance_type from nodename.
+
+        Valid NodeName format: {queue_name}-{static/dynamic}-{instance_type}-{number}
+        """
+        nodename_capture = re.match(r"^([a-z0-9\-_]+)-(static|dynamic)-([a-z0-9-]+.[a-z0-9-]+)-\d+$", nodename)
+        if not nodename_capture:
+            log.error("Invalid nodename format for node %s", nodename)
+            self.failed_nodes.append(nodename)
+            return None
+
+        return nodename_capture.group(1, 3)
```

Comment:
it is not ideal to have functions that return a different number of values based on some conditions. Also having a simple function that needs to parse nodenames also update the global state of failed_nodes makes it more complicated and harder to test. Shall we simply raise a InvalidNodenameError when a parsing failure occurs and have the caller handle it?

Reply:
Done